crates/sui-display/src/v2/lexer.rs

```diff
@@ -0,0 +1,846 @@
+
+/// Like [Lexeme] but owns the slice of source string. Useful for capturing context in an error
+/// message.
+#[derive(Debug)]
+pub(crate) struct OwnedLexeme(Token, usize, String);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) enum Token {
+    /// '@'
+    At,
+    /// ':'
+    Colon,
+    /// '::'
+    CColon,
+    /// ','
+    Comma,
+    /// '.'
+    Dot,
+    /// An identifier
+    Ident,
+    /// '<'
+    LAngle,
+    /// '{'
+    LBrace,
+    /// '['
+    LBracket,
+    /// '{{'
+    LLBrace,
+    /// '('
+    LParen,
+    /// A decimal number, optionally separated by underscores.
+    NumDec,
+    /// A hexadecimal number, prefixed with '0x' (not included in the span), optionally separated
+    /// by underscores.
+    NumHex,
+    /// '|'
+    Pipe,
+    /// '#'
+    Pound,
+    /// '>'
+    RAngle,
+    /// '}'
+    RBrace,
+    /// ']'
+    RBracket,
+    /// ')'
+    RParen,
+    /// '}}'
+    RRBrace,
+    /// Strings are surrounded by single quotes. Quotes and backslashes inside strings are escaped
+    /// with backslashes.
+    String,
+    /// A strand of text.
+    Text,
+
+    /// An unexpected byte in the input string.
+    Unexpected,
+
+    /// Whitespace around expressions.
+    Whitespace,
+}
+
+#[derive(Debug)]
+pub(crate) struct TokenSet<'t>(pub &'t [Token]);
+
+impl<'s> Lexer<'s> {
+    pub(crate) fn new(src: &'s str) -> Self {
+        Self {
+            src,
+            off: 0,
+            mode: Mode::Text,
+        }
+    }
+
+    /// Assuming the lexer is in text mode, return the next text token.
+    fn next_text_token(&mut self) -> Option<Lexeme<'s>> {
+        let bytes = self.src.as_bytes();
+
+        use Token as T;
+        Some(match bytes.first()? {
+            b'{' if bytes.get(1) == Some(&b'{') => {
+                self.advance(1);
+                self.take(T::LLBrace, 1)
+            }
+
+            b'{' => {
+                self.mode = Mode::Expr;
+                self.take(T::LBrace, 1)
+            }
+
+            b'}' if bytes.get(1) == Some(&b'}') => {
+                self.advance(1);
+                self.take(T::RRBrace, 1)
+            }
+
+            // This is not a valid token within text, but recognise it so that the parser can
+            // produce a better error message.
+            b'}' => self.take(T::RBrace, 1),
+
+            _ => self.take_until(T::Text, |b| b"{}".contains(&b)),
+        })
+    }
+
+    /// Assuming the lexer is in expression mode, return the next expression token.
+    fn next_expr_token(&mut self) -> Option<Lexeme<'s>> {
+        let bytes = self.src.as_bytes();
+
+        use Token as T;
+        Some(match bytes.first()? {
+            b'@' => self.take(T::At, 1),
+
+            b':' if bytes.get(1) == Some(&b':') => self.take(T::CColon, 2),
+
+            b':' => self.take(T::Colon, 1),
+
+            b',' => self.take(T::Comma, 1),
+
+            b'.' => self.take(T::Dot, 1),
+
+            b'0' if bytes.get(1) == Some(&b'x')
```

Comment:
Is `0x_`/`0x__` (/hex with no number to the RHS) parsed correctly/fail as expected.